### PR TITLE
Use `GetTempPath2W` if available.

### PR DESCRIPTION
**Pull request by teo-tsirpanis**
Since Windows 11 there is a new API to get the temporary files path called [`GetTempPath2`](https://docs.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-gettemppath2w). This API returns a directory inaccessible to non-SYSTEM processes if the calling process runs as SYSTEM, and [it is recommended to call this function instead of `GetTempPath`](https://docs.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-gettemppathw#:~:text=Apps%20should%20call%20GetTempPath2%20instead%20of%20GetTempPath.).

This PR tries to find `GetTempPath2W` and uses that, otherwise it falls back to `GetTempPathW`.

I also fixed a bug where the original `GetTempPathW` P/Invoke was not marked with `SetLastError` and would fail with `Operation Completed Successfuly` in case of a failure.

### CODE REVIEW

**Comment by jkotas** [comment]
Does this behavior change need to be treated as a breaking change?

File: src/libraries/System.Private.CoreLib/src/System/IO/Path.Windows.cs [hunk 1]
```diff
@@ -151,10 +154,22 @@ public static string GetTempPath()
             return path;
         }
 
+        private static unsafe delegate* unmanaged<int, char*, uint> GetGetTempPathWFunc()
+        {
+            IntPtr kernel32 = NativeLibrary.Load(Interop.Libraries.Kernel32);
```
[the remark this answers is not in the document]
Comment (jkotas) [reply src/libraries/System.Private.CoreLib/src/System/IO/Path.Windows.cs:159]: (You can also fix the one in `Architecture OSArchitecture` to do the same.)

File: src/libraries/System.Private.CoreLib/src/System/IO/Path.Windows.cs [hunk 2]
```diff
@@ -3,12 +3,15 @@
...
 using System.Diagnostics.CodeAnalysis;
+using System.Runtime.InteropServices;
 using System.Text;
 
 namespace System.IO
 {
-    public static partial class Path
+    public static unsafe partial class Path
     {
+        private static readonly delegate* unmanaged<int, char*, uint> s_GetTempPathWFunc = GetGetTempPathWFunc();
```
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/IO/Path.Windows.cs:13]: This initialization will get triggered any time other statics on Path get used. It is unnecessary and not trimmer friendly. We should initialize the pointer only when somebody actually calls GetTempPath.

Comment (teo-tsirpanis) [reply src/libraries/System.Private.CoreLib/src/System/IO/Path.Windows.cs]: Done.

File: src/libraries/System.Private.CoreLib/src/System/IO/Path.Windows.cs [hunk 3]
```diff
@@ -164,6 +179,26 @@ private static void GetTempPath(ref ValueStringBuilder builder)
                 throw Win32Marshal.GetExceptionForLastWin32Error();
 
             builder.Length = (int)result;
+
+            static uint GetTempPathW(int bufferLen, ref char buffer)
+            {
+                delegate* unmanaged<int, char*, uint> func = s_GetTempPathWFunc;
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/IO/Path.Windows.cs:185]: @jkotas, I assume we don't need to be concerned about introduced reads such that `s_GetTempPathWFunc` would need to be volatile?

Comment (jkotas) [reply src/libraries/System.Private.CoreLib/src/System/IO/Path.Windows.cs:185]: This can use `volatile` on the variable for a good measure.

It should work without it today, but I think it is a good practice to use `volatile` for anything lock-free by default.

**Comment by danmoseley** [comment]
> Does this behavior change need to be treated as a breaking change?

This would be something like -- I wrote a file to temp, upgrade my .NET or OS, and try to read it? That seems unlikely to be something we want to guarantee. In general it is good to not assume that a reboot will preserve temp.

**Comment by teo-tsirpanis** [comment]
And the path would change only if the program runs as SYSTEM. Most user apps are unaffected.

**Comment by danmoseley** [comment]
Have people tried running .NET code as SYSTEM?

**Comment by GrabYourPitchforks** [comment]
It's not uncommon for Windows services to run as SYSTEM. It's also possible that `GetTempPath2W` uses the current impersonation token instead of the process's own token, which could means the answer could change while the application is running. But I don't have a good way of testing this right now.

File: src/libraries/System.Private.CoreLib/src/System/IO/Path.Windows.cs [hunk 4]
```diff
@@ -164,6 +179,26 @@ private static void GetTempPath(ref ValueStringBuilder builder)
...
+                }
+
+                int lastError;
+                uint retVal;
+                fixed (char* ptr = &buffer)
+                {
+                    Marshal.SetLastSystemError(0);
+                    retVal = func(bufferLen, ptr);
+                    lastError = Marshal.GetLastSystemError();
+                }
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/IO/Path.Windows.cs:198]:
```suggestion

                }
```

**Review by stephentoub — APPROVED** [review]

File: src/libraries/System.Private.CoreLib/src/System/IO/Path.Windows.cs [hunk 5]
```diff
@@ -164,6 +179,27 @@ private static void GetTempPath(ref ValueStringBuilder builder)
...
+                int lastError;
+                uint retVal;
+                fixed (char* ptr = &buffer)
+                {
+                    Marshal.SetLastSystemError(0);
+                    retVal = func(bufferLen, ptr);
+                    lastError = Marshal.GetLastSystemError();
+
+                }
+                Marshal.SetLastPInvokeError(lastError);
```
Comment (Jozkee) [inline src/libraries/System.Private.CoreLib/src/System/IO/Path.Windows.cs:200]: the newline suggestion was for _after_ the brace.
```suggestion
                    lastError = Marshal.GetLastSystemError();
                }
                
                Marshal.SetLastPInvokeError(lastError);
```

File: src/libraries/System.Private.CoreLib/src/System/IO/Path.Windows.cs [hunk 6]
```diff
@@ -164,6 +179,27 @@ private static void GetTempPath(ref ValueStringBuilder builder)
...
+                if (func == null)
+                {
+                    func = s_GetTempPathWFunc = GetGetTempPathWFunc();
+                }
+
+                int lastError;
+                uint retVal;
+                fixed (char* ptr = &buffer)
+                {
+                    Marshal.SetLastSystemError(0);
```
Comment (Jozkee) [inline src/libraries/System.Private.CoreLib/src/System/IO/Path.Windows.cs:195]: Why you need to `SetLastSystemError(0)`?

Comment (teo-tsirpanis) [reply src/libraries/System.Private.CoreLib/src/System/IO/Path.Windows.cs:195]: That's what the source-generated marshalling code does.

Comment (danmoseley) [reply src/libraries/System.Private.CoreLib/src/System/IO/Path.Windows.cs:195]: Presumably to avoid random behavior if calling a function that does not set the last error, and you nevertheless read it. @elinor-fung

Comment (danmoseley) [reply src/libraries/System.Private.CoreLib/src/System/IO/Path.Windows.cs:195]: Although if e we trust the documentation of both functions, it is not necessary.

**Comment by danmoseley** [comment]
GetGCMemoryInfo failure is already resolved.

**Comment by danmoseley** [comment]
@teo-tsirpanis could you please resolve the conflict so we can merge?

**Comment by teo-tsirpanis** [comment]
Conflicts were resolved.

**Comment by danmoseley** [comment]
Failure is https://github.com/dotnet/runtime/issues/73679

File: src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/RuntimeInformation.Windows.cs [hunk 7]
```diff
@@ -42,7 +42,8 @@ public static unsafe Architecture OSArchitecture
                     // If we are running an x64 process on a non-x64 windows machine, we will report x64 as OS architecture.
                     //
                     // IsWow64Process2 is only available on Windows 10+, so we will perform run-time introspection via indirect load
-                    if (NativeLibrary.TryGetExport(NativeLibrary.Load(Interop.Libraries.Kernel32), "IsWow64Process2", out IntPtr isWow64Process2Ptr))
+                    IntPtr kernel32 = Interop.Kernel32.LoadLibraryEx(Interop.Libraries.Kernel32, 0, Interop.Kernel32.LOAD_LIBRARY_SEARCH_SYSTEM32);
```
Comment (danmoseley) [inline src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/RuntimeInformation.Windows.cs:45]: Unrelated change to make load library more secure?

Comment (jkotas) [reply src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/RuntimeInformation.Windows.cs:45]: Yes: https://github.com/dotnet/runtime/pull/72452#discussion_r924619677

**Review by adamsitnik — APPROVED** [review]
LGTM, thank you for your contribution @teo-tsirpanis !

**Comment by adamsitnik** [comment]
The failure is unrelated (#73679), merging
